shtab/__init__.py

```diff
@@ -420,10 +420,11 @@ def complete_bash(parser, root_prefix=None, preamble="", choice_functions=None):
     COMPREPLY=( $(compgen -W "${current_option_strings[*]}" -- "${completing_word}") )
   else
     # use choices & compgen
-    local IFS=$'\\n'
-    COMPREPLY=( $(compgen -W "${current_action_choices}" -- "${completing_word}") \\
-                $([ -n "${current_action_compgen}" ] \\
+    local IFS=$'\\n' # items may contain spaces, so delimit using newline
+    COMPREPLY=( $([ -n "${current_action_compgen}" ] \\
                   && "${current_action_compgen}" "${completing_word}") )
+    unset IFS
```

Comment:
@casperdcl Re [your comment on SO](https://stackoverflow.com/questions/1146098/properly-handling-spaces-and-quotes-in-bash-completion#comment131082160_11536437):

> you can also `unset IFS` as soon as it's served its purpose. That way subsequent commands can be simplified

No it's required by the `compgen -W` line for the same reason I wrote on the answer.

Instead of going over them again, I'll just give you examples showing you 1. why it's not the culprit, 2. why it's required.  After that, I'll point out 3. where the bug in this code is.

### 1. Why it's not the culprit:
```bash
$ CHOICES=(bla blue)
$ printf "[%s]\n" "${CHOICES[@]}"
[bla]
[blue]
$ compgen -W "${CHOICES[*]}" bl
bla
blue
$ IFS=$'\n'
$ compgen -W "${CHOICES[*]}" bl
bla
blue
$ unset IFS
```
👆 Having `IFS` set doesn't change the outcome.

### 2. Why it's required:
```bash
$ CHOICES=("big bla" "big blue")
$ printf "[%s]\n" "${CHOICES[@]}"
[big bla]
[big blue]
$ compgen -W "${CHOICES[*]}" "big bl" # Broken without IFS setting
$ IFS=$'\n'
$ compgen -W "${CHOICES[*]}" "big bl"
big bla
big blue
$ unset IFS
```
👆 Because `IFS` wasn't set before the first `compgen -W`, `compgen` treated the candate list as `big`, `bla`, `blue` instead of `big bla` and `big blue`.

### 3. Where the bug in this code is

It's because `current_action_choices` isn't an array.  The original *_choices variable is probably an array but the following line converts it to a scalar:

```python
local current_action_choices_var="${current_action}_choices[@]"
current_action_choices="${!current_action_choices_var}"
```

Instead you need to do:
```diff
 local current_action_choices_var="${current_action}_choices[@]"
-current_action_choices="${!current_action_choices_var}"
+current_action_choices=("${!current_action_choices_var}")
```

Here's a reproduction:
```bash
$ FOO=(a b "c d")
printf "[%s]\n" "${FOO[@]}"
[a]
[b]
[c d]
$ BAR=FOO[@]
$ echo "$BAR"
FOO[@]
$ printf "[%s]\n" "${!BAR}"
[a]
[b]
[c d]
$ BROKEN_ARRAY="${!BAR}"
$ printf "[%s]\n" "${BROKEN_ARRAY[@]}"
[a b c d]
$ PROPER_ARRAY=("${!BAR}")
$ printf "[%s]\n" "${PROPER_ARRAY[@]}"
[a]
[b]
[c d]
```

> (e.g. no need to `printf ;%q\n`)

On a final note, you also wrote 👆 in your comment but `printf %q` has nothing to do with `IFS`.

It's there to change `COMPREPLY` to `big\ bla` and `big\ blue` instead of `big bla` and `big blue`:

```bash
$ printf "%q\n" "big bla" "big blue"
big\ bla
big\ blue
```

Whether it's required or not depends entirely on the command and whether `big` and `blue` should be treated as a single argument or two separate arguments.  e.g. If "big blue" was a name of a file and auto-completion did `cat big blue` instead of `cat big\ blue` it wouldn't be very helpful.